### ISSUE TITLE
cargo vendor breaks crates such as rusty_v8, removes .gn file

### ISSUE THREAD

**Issue by apaprocki** [rust-lang/cargo#9481]
https://github.com/rust-lang/cargo/blob/070e459c2d8b79c5b2ac5218064e7603329c92ae/src/cargo/sources/path.rs#L110-L121

The denoland/rusty_v8 crate has a `.gn` file that is [included](https://github.com/denoland/rusty_v8/blob/main/Cargo.toml#L23-L60) by saying:
```
exclude = [
 ...
 ".*",
 ...
 "!.gn"
]
```

This excludes all dotfiles except `.gn`, but when this crate is vendored into a project, the `.gn` file is missing.  Given that this is legal syntax, it seems the `path.rs` logic above needs to handle this situation?

**Comment by ehuss** [comment]
This should be fixed via #9186. Can you try the latest nightly to see if it works? `cargo +nightly package --list` can be used to verify which files are included.

**Comment by weihanglo** [comment]
This snippet can also verify the exclusion rule.

```bash
git clone https://github.com/denoland/deno.git
cd deno
cargo +nightly vendor
ls vendor/rusty_v8/.gn
```

In addition, [these files](https://github.com/denoland/rusty_v8/blob/be1c2073ec1675970b4358016dc8a98ce5356f97/Cargo.toml#L62-L69) cannot be re-included due to the limitation of [gitignore rule](https://git-scm.com/docs/gitignore#_pattern_format).

> It is not possible to re-include a file if a parent directory of that file is excluded. Git doesn’t list excluded directories for performance reasons, so any patterns on contained files have no effect, no matter where they are defined.

**Comment by samanpa** [comment]
Confirmed that it works on nightly.

**Comment by PaulDance** [comment]
@apaprocki Does it work on nightly for you as well?

**Comment by apaprocki** [comment]
@PaulDance yes, nightly worked

**Comment by ehuss** [comment]
Thanks for the check, closing as fixed by #9186.
